Guide for Flax partitioning (auto SPMD) API (#2730)
google/flax: docs/guides/spmd_api.md

---
jupyter:
  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.8
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
---

<!-- #region id="2-JYevEklHfB" -->

> # Scale up Flax modules with `pjit`

Reviewer: Suggestion:

Maybe mention multi-host/multi-devices in the title so that users know what this scaling up with `pjit` means without indepth JAX knowledge. Similar to "Using JAX in multi-host and multi-process environments" https://jax.readthedocs.io/en/latest/multi_process.html

> # Scale up Flax modules with `pjit`

Author: Renamed to `flax_on_pjit.md`.

> # Scale up Flax modules with `pjit`

Reviewer 2: `flax_on_pjit.md` is better, but it still does not really match the title `Scale up Flax modules on multiple devices` ...

In this guide, you will learn how to use JAX's `pjit` and `flax.linen.spmd` to scale up your Flax modules on multiple devices and hosts.

> Modern JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [SPMD](https://jax.readthedocs.io/en/latest/glossary.html?highlight=spmd#term-SPMD) model. You can think of `pjit` as [`jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where one explicitly specificies how the input and output data should be partitioned across devices. See the [JAX guide](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) for more information on pjit.

Reviewer: Suggestions:

- Do we need "modern"? (Currently, there's just one JAX 0.4, isn't it?)
- Spell out "Single Program Multi Data" before abbreviating it.
- Use the SPMD link without the `highlight=...` part like `https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD)`
- Last sentence:
  - "You can think of `pjit` as [`jax.jit`]..." (with `jax.jit`).
  - Use "Refer to" instead of "See" (more accessibility-friendly language), "JAX guide" -> "JAX-101 `pjit` tutorial" (more precise).
  - `pjit` (formatting). 

```suggestion
JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [Single Program Multiple Device (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) model. You can think of JAX's `pjit` as [`jax.jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where you can explicitly specify how the input and output data should be partitioned across devices. Refer to the [JAX-101 `pjit` tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) for more information on `pjit`.
```

> Modern JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [SPMD](https://jax.readthedocs.io/en/latest/glossary.html?highlight=spmd#term-SPMD) model. You can think of `pjit` as [`jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where one explicitly specificies how the input and output data should be partitioned across devices. See the [JAX guide](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) for more information on pjit.

Author: Interesting. I rewrote the introductory paragraph with more basic explanation.

Flax provides an interface for you to specify partitions for your module parameters when defining your module, and later use it for `pjit` compilation. You can also use logical axis annotations to decouple your model code and partition plan, in order to customize and experiment with different partition layouts more easily.
<!-- #endregion -->

<!-- #region id="eTGIfmDclxJ2" -->
# Setup

Install Flax from head and make imports as necessary. 

Note that this Colab uses `--xla_force_host_platform_device_count=8` to emulate multiple devices on a CPU environment. You may also choose to run on a multi-device TPU environment following [this guide](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html#setup), in which case you should ignore the `os.environ` cell.
<!-- #endregion -->

```python id="e3d005e672d6"

> # TODO: replace this with `pip3 install flax` when 0.6.4 release rolls out.

Author: Since this is a public doc, it's probably better to not expose our names. I can open an issue to remind myself of that.

> # TODO: replace this with `pip3 install flax` when 0.6.4 release rolls out.

Reviewer: fyi agreed with @IvyZX 

```
# Once Flax v0.6.4 is out, replace this with `pip3 install flax`.
```

This way, the docs can be more or less up-to-date after 0.6.4.

> # TODO: replace this with `pip3 install flax` when 0.6.4 release rolls out.

Reviewer 2: You could use Github username (which is already known) ... Or create an issue and assign it to yourself.

The main motivation here is to avoid accumulating TODOs in public docs.

!pip3 install -qq "git+https://github.com/google/flax.git@main#egg=flax"
```

```python id="FJzB9vfp-OoK"
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
```

```python id="9JJbxi0zHmn3"
import functools
import numpy as np
import jax

from jax import lax, random, numpy as jnp

import flax
from flax import struct, traverse_util, linen as nn
from flax.linen import spmd
from flax.core import freeze, unfreeze
from flax.training import train_state, checkpoints

import optax
```

<!-- #region id="XfuDTqt1mO6g" -->

> Import all the `pjit` related libraries. Note that they are still in the experimental package of JAX, so some API may change. In the near future, the plan is to move the `pjit` API out of the experimental status and merge it with the current `jax.jit` API.

Author: Moving extra comments in note.

> Start a device mesh using the 8 devices available. In this example, we set them as a `2x4` device mesh, same as the layout of TPU v3-8, and annotate each axis with a name. A typical way to annotate axis names is `('data', 'model')`: `'data'` being the mesh dimension used for data-parallel sharding of the batch dimension of inputs and activations, and the `'model'` being the mesh dimension used for sharding parameters of the model across devices.

Reviewer 3: Thinking about this some more, if you have a `2x4` device and you annotate axis names with `('data', 'model')`, then this mean that if you partition over the `data` dimension you are effectively doing a two-way partition over the first dimension of the devices, and if you partition over the `model` dimension you partition it four way over the second dimension, right? It might be good to explain this explicitly, so the relation between the devices and the axis names becomes clearer.

<!-- #endregion -->

```python id="YChzss7uwwHx"
class DotReluDot(nn.Module):
  depth: int
  @nn.compact
  def __call__(self, x):
    W1 = self.param(
        'W1', 
        nn.with_partitioning(nn.initializers.xavier_normal(), (None, 'model')),
        (x.shape[-1], self.depth))

    y = jax.nn.relu(jnp.dot(x, W1))
    # force a local sharding annotation.
    y = with_sharding_constraint(y, PartitionSpec('data', 'model'))

    W2 = self.param(
        'W2', 
        nn.with_partitioning(nn.initializers.xavier_normal(), ('model', None)),
        (self.depth, x.shape[-1]))

    z = jnp.dot(y, W2)
    # force a local sharding annotation.
    z = with_sharding_constraint(z, PartitionSpec('data', None))

    # Return a tuple to conform with the API `nn.scan` as shown in the cell below.
    return z, None

```

<!-- #region id="1gvf6_a_OeU3" -->
Now define the `MLP` model as multiple layers of `DotReluDot`.

> Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.

Reviewer 3: ```suggestion
Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension for the params (the dimension over which `scan` is applied), and this dimension should be speficied using the `metadata_params` argument. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
```

> Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.

Author: I put some more concrete explanation in `metadata_params`.

To replicate identical layers, you can either use `nn.scan` or a for-loop. `nn.scan` offers faster compilation times, wheras the for-loop is faster on runtime. This guide uses `nn.scan` simply to show that lifted transforms are supported in this API.
<!-- #endregion -->

```python id="_noFQzrgJ2j0"
class MLP(nn.Module):
  num_layers: int
  depth: int
  use_scan: bool
  @nn.compact
  def __call__(self, x):
    if self.use_scan:
      x, _ = nn.scan(DotReluDot, length=self.num_layers, 
                    variable_axes={"params": 0},
                    split_rngs={"params": True},

>                     metadata_params={nn.PARTITION_NAME: None}

Author: This is just a constant to denote the partition name that annotates the `nn.scan` layer.

>                     metadata_params={nn.PARTITION_NAME: None}

Reviewer 2: it's clearer now with the added "Note:"

                    )(self.depth)(x)
    else:
      for i in range(self.num_layers):
        x, _ = DotReluDot(self.depth)(x)
    return x
```

<!-- #region id="pZ6E5btPr5yZ" -->
# Sharding specification

Now we want to generate the `PartitionSpec` that `pjit` should receive as annotations of input and output data.

The input is relatively simple. For data parallelism, we can shard the batched input `x`s across the `data` axis, by denoting the batch axis as `data`.

<!-- #endregion -->

```python id="63TwLe6vNtID"
# Input partition rule - data-parallel only
x_spec = PartitionSpec('data', None)  # dimensions: (batch, length)
x_spec
```

<!-- #region id="3d6b15d08980" -->

> However, to generate `PartitionSpec` for the output, we need to use some actual output as reference. One solution is to create a model and evaluate `model.init` abstractly using `jax.eval_shape`, and then use `nn.get_partition_spec` to automatically generate the `PartitionSpec`. 

Reviewer 3: "we need to use some actual output as reference" --> don't we need to generate the variable dict structure rather than the model outputs?

Author: That works too. I am just trying to show a more complicated example here, in which the pjitted function output is another dataclass instead of the simple variable dict. Added the explanation in doc.

The code below shows how to get the output spec if you use `flax.training.train_state` to carry out your initialization and training steps, in which case your `pjit`ted function will output a `TrainState`.

> A side note: Here we define our `init_fn` as purely functional and takes `model` and `optimizer` as arguments. This is not necessary - you can simply define with `def init_fn(k, x):` and all will work fine here.
> 

> > This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. But this does a bit of trouble because `jax.eval_shape` only takes numeric inputs, and we have to create an abstract closure before feeding the function in.

Reviewer 3: ```suggestion
> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. However, this is problematic because `jax.eval_shape` only takes numeric inputs, so we have to create an abstract closure before feeding the function in.
```

<!-- #endregion -->

```python id="YqSxbGkxo4QV"
# MLP hyperparams
B, L, E, S = 8, 4, 1024, True
# fake inputs
x = jnp.ones((B, E))
k = random.PRNGKey(0)

optimizer = optax.adam(learning_rate=0.001)
model = MLP(L, E, S)

def init_fn(k, x, model, optimizer):
  variables = model.init(k, x)
  state = train_state.TrainState.create(
    apply_fn=model.apply,
    params=variables['params'],
    tx=optimizer)
  return state

# make an abstract closure here because jax.eval_shape only takes numeric values
init_fn_abstract = functools.partial(init_fn, model=model, optimizer=optimizer)
with mesh:
  abstract_variables = jax.eval_shape(init_fn_abstract, k, x)
# we need a rule on every leaf of the input variables, in this case a nested dict.
state_spec = nn.get_partition_spec(abstract_variables)
state_spec

> ```

Reviewer 3: The output shows now three dimensions for the `PartitionSpec`, while you only provide 2 as input. I suppose this is because of the scan, but it is a but surprising so it might be worth clarifying this.

Author: The PartitionSpec shows that the params W1 and W2 has three dimensions, which WAI. The output of both pjitted functions are `TrainState`, which contains the params.

<!-- #region id="KAyPfNP9pTR4" -->

> # `pjit` the initialization and train step

Reviewer 3: Initialization with pjit is complex! Quite a number of steps, maybe we can simplify this with a simple utility function (not now, but maybe as a follow-up?)

Author: You can use decorator `@functools.partial(pjit, ...)` like `jit`. I added the explanation in the doc.

Now you can `pjit` the `init_fn` in a similar fashion as `jit`, with two extra args `in_axis_resources` and `out_axis_resources`.

You need to add `with mesh:` context when running a `pjit`ted function, so that it can refer to `mesh` to allocate data on devices correctly.
<!-- #endregion -->

```python id="RFxvZjwPS5Za"
pjit_init_fn = pjit(init_fn,
                    static_argnums=(2, 3),
                    in_axis_resources=(PartitionSpec(None), x_spec),  # RNG key and x

>                     out_axis_resources=state_spec

Author: JAX team is making `pjit` more automatic so that user don't have to specify input and output axis. But as far as Anselm and I know, this approach is not very efficient on complex model for now, so we decided to still present the version that explicitly states the axis.

>                     out_axis_resources=state_spec

Reviewer 2: ah ... so it's still a bit rough around the edges, I see.

I think I would still mention this because in many examples users will see examples of `pjit()` where the `out_axis_resources` is omitted (e.g. our very own `flax.core.meta.Partitioned`).

                    )
with mesh:
  initialized_state = pjit_init_fn(k, x, model, optimizer)
jax.tree_map(jnp.shape, initialized_state)
```

<!-- #region id="qgVa3lqLgPLz" -->
## Looking into the module output

Note that in the output `initialized_state`, the params `W1` and `W2` are of type [`Partitioned`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Partitioned.html). This is a wrapper around the actual JAX array that allows Flax to record metadata associated with it. You can access the raw JAX array by adding `.value` or running `.unbox()`.

> You can also check the underlying `.sharding` of the JAX array, which gives a hint on the way it is partitioned.

Author: This is a fairly new concept in JAX and is not very well documented as for now. I only find one mentioning of OpSharding here: https://jax.readthedocs.io/en/latest/jax_array_migration.html#why-create-jax-array

<!-- #endregion -->

```python id="UEj1A2BhhUcC"
print(type(initialized_state.params['ScanDotReluDot_0']['W1']))
print(type(initialized_state.params['ScanDotReluDot_0']['W1'].value))
print(initialized_state.params['ScanDotReluDot_0']['W1'].value.shape)
```

```python id="-Hhp_glBOL1R"
print(initialized_state.params['ScanDotReluDot_0']['W1'].value.sharding)
```

<!-- #region id="Yu7ohrBzN-eP" -->
You can use `jax.tree_map` to perform mass computation on a dict of boxed params, in the same way as on a dict of JAX arrays.
<!-- #endregion -->

```python id="aqwt7F0Kmn0f"
diff = jax.tree_map(
    lambda a, b: a - b, 
    initialized_state.params['ScanDotReluDot_0'], initialized_state.params['ScanDotReluDot_0'])
print(jax.tree_map(jnp.shape, diff))
diff_array = diff['W1'].unbox()
print(type(diff_array))
print(diff_array.shape)
```

<!-- #region id="dcLFwH1dhr3v" -->

> ## `pjit` the train step and inference 

Reviewer 3: It isn't clear which parts of these code blocks refer to "train step" and which refer to "inference". Maybe add some text before the inference code block to clarify that.

Now we can do the same and make a `pjit`ted training step.
<!-- #endregion -->

```python id="KF5lmJ5FwxqJ"
def train_step(state, x):
  # fake loss fn
  def loss_unrolled(params):
    y = model.apply({'params': params}, x)
    return y.sum()
  grad_fn = jax.grad(loss_unrolled)
  grads = grad_fn(state.params)
  state = state.apply_gradients(grads=grads)
  return state

pjit_step_fn = pjit(train_step,
                    (state_spec, x_spec),  # input annotations
                    state_spec,            # output annotations
                    )
```

```python id="F5gy_gb_mOqR"

> with mesh:

Reviewer 2: nit: in the other examples, the `with mesh:` code is in the same cell as the function being pjitted

  new_state = pjit_step_fn(initialized_state, x)
```

```python id="0BE_qTV-vbfo"
def apply(state, x):
  return state.apply_fn({'params': state.params}, x)

pjit_apply_fn = pjit(apply, (state_spec, x_spec), x_spec)

with mesh:
  y = pjit_apply_fn(new_state, x)
print(type(y))
print(y.dtype)
print(y.shape)
```

<!-- #region id="vXY2mx7WjOyv" -->
## Profiling

> If you are running on a TPU pod or pod slice, you can use the `block_all` util below to profile the cell below to observe a performance increase.

Reviewer 3: ```suggestion
If you are running on a TPU pod or pod slice, you can use the `block_all` utility function below to observe a performance increase.
```

> If you are running on a TPU pod or pod slice, you can use the `block_all` util below to profile the cell below to observe a performance increase.

Reviewer 2: how do I see the "performance increase" ?

<!-- #endregion -->

```python id="A2Jn7kktw_UZ"
%%timeit

# check this out by profiling, your model is made model parallel automatically by XLA:SPMD!
def block_all(xs):
  jax.tree_map(lambda x: x.block_until_ready(), xs)
  return xs

with mesh:
  new_state = block_all(pjit_step_fn(initialized_state, x))
```

<!-- #region id="GUXkJiQkotfm" -->
# Logical axis annotation

> JAX auto SPMD encourages user to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 

Reviewer 3: ```suggestion
JAX auto SPMD encourages users to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 
```

Check out the `Logical-` model definitions below. It's exactly the same with the model above, except for two differences:

> 1. Each axis are annotated with more concrete, meaningful names - like `embed`, `hidden`, `batch` and `layer`. These names are referred as "logical axis names" in Flax. They make the dimensional changes inside model definition more readable.

Reviewer 3: ```suggestion
1. All axes are annotated with more concrete, meaningful names - like `embed`, `hidden`, `batch` and `layer`. These names are referred as "logical axis names" in Flax. They make the dimensional changes inside model definitions more readable.
```

from jax.experimental.pjit import pjit, with_sharding_constraint, PartitionSpec
from jax.experimental.maps import Mesh
from jax.experimental import mesh_utils

device_mesh = mesh_utils.create_device_mesh((2, 4))
print(device_mesh)
mesh = Mesh(devices=device_mesh, axis_names=('data', 'model'))
mesh
```

<!-- #region id="SJcTXjGYhe5o" -->
# Model Definition

Define an example layer `DotReluDot`, which creates two parameters `W1` and `W2` for dot products, and does a `jax.nn.relu` in between. To use it in `pjit` efficiently, you can use APIs in `pjit` and `flax.core.meta` to annotate the parameters and intermediate variables correctly:

* Use `meta.with_partitioning` to decorate the initializer function when creating params `W1` and `W2`.

* Use `pjit.with_sharding_constraint` to annotate intermediate variables like `y` and `z` to force a particular sharding pattern under pjit when the ideal constraint is known.

 * This is optional, but sometimes can help auto SPMD to partition efficiently.

> Each value passed into a sharding annotates a dimension, and its value should be among `data`, `model` or `None`. This refers to whether the data's axis should be sharded across one of the device mesh dimensions, or not sharded at all.

Author: Yes, this is correct. I rephrased it to:

Calling these APIs annotates each dimension of your parameter and intermediate variables with an axis name, namely `'data'`, `'model'` or `None`. This refers to how this dimension should be sharded - across one of the device mesh dimensions, or not sharded at all.

For example, if we define `W1`, which has shape `(x.shape[-1], self.depth)`, with `(None, 'model')`, the first dimension (of length `x.shape[-1]`) will be replicated across all devices, while the second dimension (of length `self.depth`) will be sharded over the `model` axis of the device mesh.

<!-- #endregion -->

```python executionInfo={"elapsed": 55, "status": "ok", "timestamp": 1671158221167, "user": {"displayName": "Ivy Zheng", "userId": "15297372265856137303"}, "user_tz": 480} id="YChzss7uwwHx"
class DotReluDot(nn.Module):
  depth: int
  @nn.compact
  def __call__(self, x):
    W1 = self.param(
        'W1', 
        meta.with_partitioning(
            nn.initializers.xavier_normal(),
            (None, 'model')),
        (x.shape[-1], self.depth))

    y = jax.nn.relu(jnp.dot(x, W1))

>     # force a local sharding annotation, fairly redundant in this case.

Author: The local sharding annotation here. I removed the comment here and added more explanation on the documentation above. Basically `with_sharding_constraint` is only to ensure that `pjit` shards this variable in this way, and it's redundant here because `pjit` will by default shard it this way too.

>     # force a local sharding annotation, fairly redundant in this case.

Reviewer 3: Ah makes sense, thanks for updating it!

    y = with_sharding_constraint(y, PartitionSpec('data', 'model'))

    W2 = self.param(
        'W2', 
        meta.with_partitioning(
            nn.initializers.xavier_normal(),
            ('model', None)),
        (self.depth, x.shape[-1]))

    z = jnp.dot(y, W2)
    # force a local sharding annotation, fairly redundant in this case.

>     z = with_sharding_constraint(z, PartitionSpec('data', None))

Author: Because the output should be sharded as ('data', None), in which 'data' refers to the batch dimension of input.
